File: edge-layer/edge-vision-inference/app/seed.py
#!/usr/bin/env python3
"""
Seed script for edge-vision-inference service
Creates mock inference results data
"""

import os
import sys
import asyncio
from datetime import datetime, timedelta
from decimal import Decimal

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from app.db import InferenceDb

# Guard: prevent seed in production
if os.getenv('NODE_ENV') == 'production' and not os.getenv('ALLOW_SEED_IN_PROD'):
    print('ERROR: Seed is not allowed in production!', file=sys.stderr)
    print('Set ALLOW_SEED_IN_PROD=true if you really want to seed production.', file=sys.stderr)
    sys.exit(1)

SEED_COUNT = int(os.getenv('SEED_COUNT', '30'))

# Tenant-friendly IDs (multi-tenant ready)
TENANT_1 = 't-001'
TENANT_2 = 't-002'
FARM_1A = 'f-001'
FARM_1B = 'f-002'
FARM_2A = 'f-101'
BARN_1A_1 = 'b-001'
BARN_1A_2 = 'b-002'
BARN_1B_1 = 'b-003'
DEVICE_WEIGH_1 = 'wv-001'
DEVICE_WEIGH_2 = 'wv-002'

def get_session_ids():
    """Get fixed session IDs"""
    return [f'00000000-0000-4000-8000-0000004000{(i+1):02x}' for i in range(30)]


async def seed_inference(db: InferenceDb):
    """Seed inference results"""
    print(f'Starting seed (SEED_COUNT={SEED_COUNT})...')

    await db.connect()
    await db.ensure_schema()

    session_ids = get_session_ids()
    now = datetime.utcnow()
    model_version = os.getenv('MODEL_VERSION', 'v1.0.0')

    result_count = max(SEED_COUNT, 30)
    results_created = 0

    configs = [
        {'tenant_id': TENANT_1, 'farm_id': FARM_1A, 'barn_id': BARN_1A_1, 'device_id': DEVICE_WEIGH_1},
        {'tenant_id': TENANT_1, 'farm_id': FARM_1A, 'barn_id': BARN_1A_2, 'device_id': DEVICE_WEIGH_2},
        {'tenant_id': TENANT_1, 'farm_id': FARM_1B, 'barn_id': BARN_1B_1, 'device_id': DEVICE_WEIGH_1},
        {'tenant_id': TENANT_2, 'farm_id': FARM_2A, 'barn_id': BARN_1A_1, 'device_id': DEVICE_WEIGH_2},
    ]

    for i in range(result_count):
        config = configs[i % len(configs)]
        session_id = session_ids[i % len(session_ids)]
        hours_ago = i // 5

        predicted_weight = 1.2 + (i % 20) * 0.1
        confidence = 0.85 + (i % 15) * 0.01
        if confidence > 1.0:
            confidence = 1.0

        metadata = {
            'source': 'seed',
            'index': i,
            'batch_id': f'batch-{(i % 8) + 1}',
        }

        try:
            result_id = f'00000000-0000-4000-8000-{(50000 + i):012x}'
            result_id = await db.create_inference_result(
                result_id=result_id,
                tenant_id=config['tenant_id'],
                farm_id=config['farm_id'],
                barn_id=config['barn_id'],
                device_id=config['device_id'],
                session_id=session_id,
                media_id=f'media-{i:06d}',
                predicted_weight_kg=predicted_weight,
                confidence=confidence,
                model_version=model_version,
                metadata=metadata,
            )
            results_created += 1
        except Exception as e:
            # Ignore duplicates (if any unique constraints exist)
            if 'unique' not in str(e).lower() and 'duplicate' not in str(e).lower():
                print(f'Warning: Failed to insert inference result {i}: {e}')

    print(f'Inserted {results_created} inference_results')
    if results_created < 30:
        raise RuntimeError(f'Seed expected >= 30 inference_results, got {results_created}')

    await db.close()
    print('Seed completed successfully!')
    print(f'Summary: {results_created} inference_results')


async def main():
    database_url = os.getenv(
        'DATABASE_URL',
        'postgresql://farmiq:farmiq_dev@localhost:5141/farmiq'
    )
    db = InferenceDb(database_url)
    try:
        await seed_inference(db)
    except Exception as e:
        print(f'Error during seed: {e}', file=sys.stderr)
        import traceback
        traceback.print_exc()
        sys.exit(1)


if __name__ == '__main__':
    asyncio.run(main())

File: edge-layer/edge-media-store/prisma/seed.ts
import { PrismaClient, Prisma } from '@prisma/client'
import { ensureMediaSchema } from '../src/db/ensureSchema'

const prisma = new PrismaClient()

async function main() {
  // Guard: prevent seed in production
  if (process.env.NODE_ENV === 'production' && !process.env.ALLOW_SEED_IN_PROD) {
    console.error('ERROR: Seed is not allowed in production!')
    console.error('Set ALLOW_SEED_IN_PROD=true if you really want to seed production.')
    process.exit(1)
  }

  const SEED_COUNT = parseInt(process.env.SEED_COUNT || '60', 10)
  const count = Math.max(30, SEED_COUNT)
  const tenants = ['t-001', 't-002']
  const now = Date.now()

  await prisma.$connect()
  await ensureMediaSchema(prisma)

  const rows = Array.from({ length: count }, (_, i) => {
    const tenantId = tenants[i % tenants.length]
    const farmId = tenantId === 't-001' ? 'f-001' : 'f-101'
    const barnId = tenantId === 't-001' ? 'b-001' : 'b-101'
    const deviceId = tenantId === 't-001' ? 'd-001' : 'd-101'
    const sessionId = i % 4 === 0 ? `session-${(i / 4).toString().padStart(4, '0')}` : null
    const day = ((i % 28) + 1).toString().padStart(2, '0')

    return {
      tenantId,
      farmId,
      barnId,
      deviceId,
      sessionId,
      bucket: 'farmiq-media',
      objectKey: `tenants/${tenantId}/farms/${farmId}/barns/${barnId}/devices/${deviceId}/images/2025-12-${day}/img-${i
        .toString()
        .padStart(6, '0')}.jpg`,
      etag: `etag-${i.toString().padStart(6, '0')}`,
      mimeType: 'image/jpeg',
      sizeBytes: 1024 + i,
      capturedAt: new Date(now - i * 60_000),
    }
  })

  const values = rows.map((r) =>
    Prisma.sql`(${r.tenantId}, ${r.farmId}, ${r.barnId}, ${r.deviceId}, ${r.sessionId}, ${r.bucket}, ${r.objectKey}, ${r.etag}, ${r.mimeType}, ${r.sizeBytes}, ${r.capturedAt})`
  )

  await prisma.$executeRaw(
    Prisma.sql`
      INSERT INTO media_objects
        (tenant_id, farm_id, barn_id, device_id, session_id, bucket, object_key, etag, mime_type, size_bytes, captured_at)
      VALUES
        ${Prisma.join(values)}
      ON CONFLICT (tenant_id, object_key)
      DO UPDATE SET
        farm_id = EXCLUDED.farm_id,
        barn_id = EXCLUDED.barn_id,
        device_id = EXCLUDED.device_id,
        session_id = EXCLUDED.session_id,
        bucket = EXCLUDED.bucket,
        etag = EXCLUDED.etag,
        mime_type = EXCLUDED.mime_type,
        size_bytes = EXCLUDED.size_bytes,
        captured_at = EXCLUDED.captured_at
    `
  )

  console.log(`Upserted ${rows.length} media_objects rows`)
}

main()
  .catch((error) => {
    console.error('Seed failed', error)
    process.exitCode = 1
  })
  .finally(async () => {
    await prisma.$disconnect()
  })
